src/jit/lclvars.cpp

```diff
@@ -2706,54 +2689,60 @@ void Compiler::lvaUpdateClass(unsigned varNum, CORINFO_CLASS_HANDLE clsHnd, bool
     // We should already have a class
     assert(varDsc->lvClassHnd != nullptr);
 
-#if defined(DEBUG)
+    // We should only be updating classes for single-def locals.
+    assert(varDsc->lvSingleDef);
 
-    // In general we only expect one update per local var. However if
-    // a block is re-imported and that block has the only STLOC for
-    // the var, we may see multiple updates. All subsequent updates
-    // should agree on the type, since reimportation is triggered by
-    // type mismatches for things other than ref types.
-    if (varDsc->lvClassInfoUpdated)
-    {
-        assert(varDsc->lvClassHnd == clsHnd);
-        assert(varDsc->lvClassIsExact == isExact);
-    }
+    // Now see if we should update.
+    //
+    // New information may not always be "better" so do some
+    // simple analysis to decide if the update is worthwhile.
+    const bool isNewClass   = (clsHnd != varDsc->lvClassHnd);
+    bool       shouldUpdate = false;
 
-    // This counts as an update, even if nothing changes.
-    varDsc->lvClassInfoUpdated = true;
+    // Are we attempting to update the class? Only check this when we have
+    // an new type and the existing class is inexact... we should not be
+    // updating exact classes.
+    if (!varDsc->lvClassIsExact && isNewClass)
+    {
+        // Todo: improve this analysis by adding a new jit interface method
+        DWORD newAttrs = info.compCompHnd->getClassAttribs(clsHnd);
```

Comment:
Computing class attribs is relatively expensive operation. It computes a ton more than just the `CORINFO_FLG_SHAREDINST` flag that you care about here. It would be better to change `mergeClasses` to return NULL when any of the types is shared.